design/baremetal-operator/bulk-set-bios-config.md

```diff
@@ -0,0 +1,228 @@
+manner across all similar hardware. Create a template for the current
+BIOS configurations settings cached on a server. Give access to the BIOS
+registry which can be used as documentation so a user can easily update
+the BIOS settings to desired values. Validate the BIOS Configuration
+values based on the vendor's BIOS Registry before sending to the BMC.
+
+### Non-Goals
+
+This proposal does not attempt to fulfill the following goals:
+
+- Allow the transfer of BIOS configuration between different vendor’s
+  servers or different vendor models
+- Support bulk set of other configurations besides BIOS
+- Provide storage mechanism for BIOS settings
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+When provisioning a large amount of machines with identical settings, our
+customers want to provide a known, validated payload with all the desired
+firmware configuration.
+
+This ensures all the machines will have the identical settings, reducing issues
+caused by unexpected configuration differences.
+
+Since the name/value pair mapping is vendor specific and not intuitive, it
+doesn’t lend itself to creating the input by hand using existing vendor
+documentation, except for potentially a small subset of names. It’s desirable
+to take an existing BIOS config set and modify the current settings to the
+desired settings. The BIOS Registry can be used as a source of documentation
+to modify the settings.
+
+## Design Details
+
+This proposes a new Custom Resource Definition (CRD) to store the settings
+read from Ironic along with the modified settings to write to Ironic. Since
+each vendor provides 100 or more BIOS Settings, its not practical to store
+these in the BareMetalHost CRD. The new CRD will be named
+`HostFirmwareSettings` and will consist of the following:
+
+- `currentFirmwareSettings` - the current BIOS Settings retrieved from Ironic
+  via the bios API will be stored in the Status section. This dataset includes
+  BIOS Registry information which provides a schema for the settings.
+- `writeableFirmwareSettings` - BIOS Settings to write to the host via Ironic
+  will be stored in the Spec section. If it is empty, it will be initially
+  populated with the name/value pairs from `currentFirmwareSettings` that do
+  not have the `read-only` or `unique` flag set.
+
+The BIOS Settings are retrieved from the BMC by Ironic and cached
+whenever the node moves to `manageable` or `cleaning`, or when the settings
+are updated. Recent changes to Ironic (see the Ironic proposal
+[BIOS Registry](https://review.opendev.org/c/openstack/ironic-specs/+/774681)
+added an additional query of the BIOS Registry.
+
+The baremetal-operator can get the BIOS Settings from Ironic after the node
+has been moved to manageable via the `v1/nodes/{node_ident}/bios?detail=True`
+API.  This API provides a list of BIOS Settings (aka `Attributes`) in JSON and
+each setting includes the following data:
+
+- Name
```

Comment:
For the sake of clarity, the `Name` will be vendor-specific

Reply:
Sure I can an example of the CRD here, especially now that https://github.com/metal3-io/baremetal-operator/pull/901 is posted. I'll get agreement here and update 901 with the results.

And yes the name will be vendor specific.

Reply:
Updated spec to add CRD and reduce naming. Also indicated that the name is vendor specifc.